src/test/java/engineering/swat/watch/SingleFileTests.java

```diff
@@ -117,4 +120,74 @@ void singleFileThatMonitorsOnlyADirectory() throws IOException, InterruptedExcep
                 .untilTrue(seen);
         }
     }
+
+    @Test
+    void noRescanOnOverflow() throws IOException, InterruptedException {
+        var whichFiles = OnOverflow.NONE;
+        var bookkeeper = new Bookkeeper();
+        try (var watch = startWatchAndTriggerOverflow(whichFiles, bookkeeper)) {
+            Thread.sleep(TestHelper.SHORT_WAIT.toMillis());
+
+            await("Overflow shouldn't trigger created events")
+                .untilFalse(bookkeeper.seenCreated);
+            await("Overflow shouldn't trigger modified events")
+                .untilFalse(bookkeeper.seenModified);
+            await("Overflow should be visible to user-defined event handler")
+                .untilTrue(bookkeeper.seenOverflow);
+        }
+    }
+
+    @Test
+    void memorylessRescanOnOverflow() throws IOException, InterruptedException {
+        var whichFiles = OnOverflow.ALL;
+        var bookkeeper = new Bookkeeper();
+        try (var watch = startWatchAndTriggerOverflow(whichFiles, bookkeeper)) {
+            Thread.sleep(TestHelper.SHORT_WAIT.toMillis());
+
+            await("Overflow should trigger created event")
+                .untilTrue(bookkeeper.seenCreated);
+            await("Overflow shouldn't trigger modified event (empty file)")
+                .untilFalse(bookkeeper.seenModified);
+            await("Overflow should be visible to user-defined event handler")
+                .untilTrue(bookkeeper.seenOverflow);
+        }
+    }
+
+    private ActiveWatch startWatchAndTriggerOverflow(OnOverflow whichFiles, Bookkeeper bookkeeper) throws IOException {
+        var parent = testDir.getTestDirectory();
+        var file = parent.resolve("a.txt");
+
+        var watch = Watcher
+            .watch(file, WatchScope.PATH_ONLY)
+            .approximate(whichFiles)
+            .on(bookkeeper)
+            .start();
+
+        var overflow = new WatchEvent(WatchEvent.Kind.OVERFLOW, parent);
+        ((EventHandlingWatch) watch).handleEvent(overflow);
+        return watch;
+    }
+
+    private static class Bookkeeper implements Consumer<WatchEvent> {
```

Comment:
I think we should also check we don't get any events (similated or not) for entries not the file we requested.

To make sure we're not generating a full directory list of simulated events.